File: Python/problem031.py
#! /usr/bin/env python3
"""
Problem 31 - Project Euler
http://projecteuler.net/index.php?section=problems&id=031
"""


def coin_patterns(rest, coins, pattern, ways):
    if len(coins) == 1 and coins[0] == 1:
        pattern.append(rest)
        ways.append(pattern)
    else:
        n = rest // coins[0]
        for i in range(0, n + 1):
            ways = coin_patterns(rest - coins[0] * i, coins[1:], pattern + [i], ways)
    return ways


if __name__ == "__main__":
    coins = [200, 100, 50, 20, 10, 5, 2, 1]
    ways = coin_patterns(200, coins, [], [])
    print(len(ways))

File: Clojure/problem031.clj
#! /usr/bin/env clojure
; Problem 31 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=031

(defn coin-patterns [restcoins coins pattern]
  (if (and (= (count coins) 1)
           (= (first coins) 1))
    (-> (conj pattern restcoins)
        (list))
    (->> (let [n (quot restcoins (first coins))]
           (for [i (range 0 (inc n))]
             (coin-patterns (- restcoins
                               (* i (first coins)))
                            (rest coins)
                            (conj pattern i))))
         (reduce concat))))

(def coins '(200 100 50 20 10 5 2 1))
(-> (coin-patterns 200 coins '())
    (as-> ans
          (println (count ans))))
